File: packages/contracts/python/src/cuvoy_contracts/itinerary.py
"""Itinerary, stops, routes, reservations (PROJECT_SPEC §5, §9, §14)."""

from datetime import date

from pydantic import Field

from cuvoy_contracts.common import ContractModel, CostAmount, LocalDateTime
from cuvoy_contracts.enrichment import CrowdConfidence, Weather
from cuvoy_contracts.enums import ItineraryItemType, TransportMode, WarningCode
from cuvoy_contracts.place import Place


class ReservationInfo(ContractModel):
    likely_needed: bool
    website: str | None = None
    phone: str | None = None
    guidance: str


class RouteLeg(ContractModel):
    from_place_id: str
    to_place_id: str
    duration_seconds: int
    duration_buffered_seconds: int
    distance_meters: int
    mode: TransportMode
    geometry: str | None = None
    cost: CostAmount | None = None


class ItineraryItem(ContractModel):
    type: ItineraryItemType
    start: LocalDateTime
    end: LocalDateTime
    place: Place | None = None
    title: str | None = None
    dwell_minutes: int | None = None
    wait_minutes: int | None = None
    travel_minutes: int | None = None
    travel_minutes_buffered: int | None = None
    cost: CostAmount | None = None
    crowd: CrowdConfidence | None = None
    reason: str | None = None
    warnings: list[WarningCode] = Field(default_factory=list)
    reservation: ReservationInfo | None = None
    locked: bool = False
    route: RouteLeg | None = None


class DailyCost(ContractModel):
    currency: str
    activities: CostAmount | None = None
    meals: CostAmount | None = None
    transport: CostAmount | None = None
    total_excluding_transport: float | None = None
    total_including_transport: float | None = None
    transport_shown: bool = False


class ItineraryDay(ContractModel):
    day_index: int
    date: date
    timezone: str
    city: str | None = None
    is_travel_day: bool = False
    items: list[ItineraryItem]
    daily_cost: DailyCost | None = None
    weather: Weather | None = None


class Itinerary(ContractModel):
    days: list[ItineraryDay]
    timezone: str
    currency: str
    narrative: str | None = None

File: packages/contracts/typescript/src/itinerary.ts
import { z } from "zod";

import { CostAmountSchema, LocalDateTimeSchema } from "./common";
import { CrowdConfidenceSchema, WeatherSchema } from "./enrichment";
import { ItineraryItemTypeSchema, TransportModeSchema, WarningCodeSchema } from "./enums";
import { PlaceSchema } from "./place";

export const ReservationInfoSchema = z.object({
  likely_needed: z.boolean(),
  website: z.string().optional().nullable(),
  phone: z.string().optional().nullable(),
  guidance: z.string(),
});

export const RouteLegSchema = z.object({
  from_place_id: z.string(),
  to_place_id: z.string(),
  duration_seconds: z.number().int(),
  duration_buffered_seconds: z.number().int(),
  distance_meters: z.number().int(),
  mode: TransportModeSchema,
  geometry: z.string().optional().nullable(),
  cost: CostAmountSchema.optional().nullable(),
});

export const ItineraryItemSchema = z.object({
  type: ItineraryItemTypeSchema,
  start: LocalDateTimeSchema,
  end: LocalDateTimeSchema,
  place: PlaceSchema.optional().nullable(),
  title: z.string().optional().nullable(),
  dwell_minutes: z.number().int().optional().nullable(),
  wait_minutes: z.number().int().optional().nullable(),
  travel_minutes: z.number().int().optional().nullable(),
  travel_minutes_buffered: z.number().int().optional().nullable(),
  cost: CostAmountSchema.optional().nullable(),
  crowd: CrowdConfidenceSchema.optional().nullable(),
  reason: z.string().optional().nullable(),
  warnings: z.array(WarningCodeSchema).default([]),
  reservation: ReservationInfoSchema.optional().nullable(),
  locked: z.boolean().default(false),
  route: RouteLegSchema.optional().nullable(),
});

export const DailyCostSchema = z.object({
  currency: z.string(),
  activities: CostAmountSchema.optional().nullable(),
  meals: CostAmountSchema.optional().nullable(),
  transport: CostAmountSchema.optional().nullable(),
  total_excluding_transport: z.number().optional().nullable(),
  total_including_transport: z.number().optional().nullable(),
  transport_shown: z.boolean().default(false),
});

export const ItineraryDaySchema = z.object({
  day_index: z.number().int(),
  date: z.string().date(),
  timezone: z.string(),
  city: z.string().optional().nullable(),
  is_travel_day: z.boolean().default(false),
  items: z.array(ItineraryItemSchema),
  daily_cost: DailyCostSchema.optional().nullable(),
  weather: WeatherSchema.optional().nullable(),
});

export const ItinerarySchema = z.object({
  days: z.array(ItineraryDaySchema),
  timezone: z.string(),
  currency: z.string(),
  narrative: z.string().optional().nullable(),
});

export type ReservationInfo = z.infer<typeof ReservationInfoSchema>;
export type RouteLeg = z.infer<typeof RouteLegSchema>;
export type ItineraryItem = z.infer<typeof ItineraryItemSchema>;
export type ItineraryDay = z.infer<typeof ItineraryDaySchema>;
export type Itinerary = z.infer<typeof ItinerarySchema>;
